### ISSUE TITLE
Packages getting mis-resolved in studio?

### ISSUE THREAD

**Issue by christophermaier** [habitat-sh/habitat#3403]
It appears that the studio build process can get confused when resolving dependencies. A community user brought this to light today with a plan that depended on both `core/curl` and `core/openssl`. The relevant bits of the dependency graph are as follows:

```
ryoseph/moodle-php5/5.6.23/20170929133635
    ...
    core/curl/7.51.0/20170513235716
        core/cacerts/2017.01.18/20170513215104
        core/glibc/2.22/20170513201042
            core/linux-headers/4.3/20170513200956
        core/openssl/1.0.2j/20170513215106 (*)
            core/glibc/2.22/20170513201042
                core/linux-headers/4.3/20170513200956
            core/zlib/1.2.8/20170513201911
                core/glibc/2.22/20170513201042
                    core/linux-headers/4.3/20170513200956
            core/cacerts/2017.01.18/20170513215104
        core/zlib/1.2.8/20170513201911
            core/glibc/2.22/20170513201042
                core/linux-headers/4.3/20170513200956
    ...
    core/openssl/1.0.2j/20170926165227 (*)
        core/glibc/2.22/20170513201042
            core/linux-headers/4.3/20170513200956
        core/zlib/1.2.8/20170513201911
            core/glibc/2.22/20170513201042
                core/linux-headers/4.3/20170513200956
        core/cacerts/2017.01.18/20170513215104
```

The user was pulling updates from the stable channel. Note that at the time this occurred, `core/openssl/1.0.2j/20170513215106` (the `core/curl` dependency) was stable, while `core/openssl/1.0.2j/20170926165227` was unstable.

The `plan.sh` only refers to `core/curl` and `core/openssl`, with no further qualifications, and `HAB_BLDR_CHANNEL` was not set.

Looking at the log output of the build, however, revealed this:

```
» Installing core/openssl from channel 'stable'
→ Using core/openssl/1.0.2j/20170513215106
★ Install of core/openssl/1.0.2j/20170513215106 complete with 0 new packages installed.
   moodle-php5: Resolved dependency 'core/openssl' to /hab/pkgs/core/openssl/1.0.2j/20170926165227
```

Note that it is installing the correct (stable) version of `core/openssl`, but it resolves the dependency to the incorrect (unstable) version.

Digging further revealed that the unstable `core/openssl` was getting on disk from the supervisor! When entering the studio, a supervisor is started by default. This results in additional supervisor dependencies being installed; at the time this occurred, it pulled in the latest unstable `core/hab-launcher` -- `core/hab-launcher/5187/20170926171711` -- which has a dependency on `/hab/pkgs/core/openssl/1.0.2j/20170926165227`. This is how the dependency came to be present on the system.

To build successfully, you can either remove the offending `core/openssl` artifacts and build, or enter the studio with `HAB_STUDIO_SUP=false` (strangely, `HAB_STUDIO_SUP='--channel=stable'` does not work; the same artifacts are pulled in.)

Thus it appears that there is a bug in dependency resolution, whereby a more recent build seems to take precedence over the one that was downloaded from Builder.

This behavior was observed running `hab 0.33.2/20170925045915` on Linux.

**Comment by fnichol** [comment]
Unfortunately, at least up to this point, this issue is a result of how the dependency revolving works by design. The build program uses an internal function [`_latest_installed_package()`](https://github.com/habitat-sh/habitat/blob/640e1c27f4e2907c665bd22d75d48b63d7f4d706/components/plan-build/bin/hab-plan-build.sh#L627-L666) which looks for the most recent installed package on disk. This implementation pre-dated even the idea of channels or running a Supervisor in the same install space as the build environment. The reason for this implementation was to allow for building multiple offline packages (i.e. no Builder hosting packages) and using them to build additional packages. In fact, the initial bootstrapping uses this to build the critical chain of base packages.

In my mind, there are a few ways to mitigate/resolve what happened here:

* Always keep in mind that `hab pkg build` will *not* install and run a Supervisor, nor will `hab studio run`. The most correct builds will start from a minimal and "pure" Studio with no additional software to taint the build.
* Take a hard look at automatically installing and starting a Supervisor in the background on `hab studio enter`. Given our development experience, we aren't going to eliminate this as a feature, but perhaps we could delay the execution of that behavior until a user asked for it, typed a command, or invoked a Supervisor subcommand. Think of it as lazy-starting the Supervisor. That way, the additional software wouldn't be installed until after a package was built *and* a user wished to test it right there.
* Statically compile the Launcher and Supervisor so that they introduce no further dependencies into the system. This has been on my hit list for 6+ months and in light of what you've found here, this could be one more concrete benefit.
* Investigate an alternate way to run the Supervisor in a Studio. At the moment, nothing is coming to my mind as all Habitat packages are absolutely linked and the Supervisor needs direct access to the same `pkg_path` as the services that are being installed and started.
* Investigate an alternative strategy to the current implementation of `_latest_installed_package()`, keeping in mind that we still need an offline mechanism to build packages against themselves without relying on a remote Builder.
* Get a lot better about rebuilding our core packages, promoting to stable, and using that stable set to build unstable CLIs, Supervisors, and Launchers. I have a feeling  this is going to happen no matter what

That's a super interesting problem--good sleuthing too to track it down!

**Comment by eeyun** [comment]
So I'm running into this today myself as will basically anyone using the studio to build a package that requires openssl as a top level `pkg_dep`. 

The digging I did looked to actually be caused not by the supervisor but by the launcher. 
In a fresh studio you can see a couple things:


```
[5][default:/src:0]# cat /hab/pkgs/core/hab-launcher/6083/20171101045646/DEPS
core/gcc-libs/5.2.0/20170513212920
core/glibc/2.22/20170513201042
core/libarchive/3.3.2/20171101045440
core/libsodium/1.0.13/20170905223149
core/openssl/1.0.2l/20171101045054
[5][default:/src:0]# cat /hab/pkgs/core/hab-sup/0.38.0/20171026221910/DEPS 
core/busybox-static/1.24.2/20170513215502
core/gcc-libs/5.2.0/20170513212920
core/glibc/2.22/20170513201042
core/libarchive/3.3.2/20171018164107
core/libsodium/1.0.13/20170905223149
core/openssl/1.0.2l/20171014213633
core/zeromq/4.2.2/20171018132502
```
So first it looks like the supervisor is using an appropriate version of openssl by default today. But, launcher version 6083 is unstable being used by default in the studio. The other thing being that you can see is that the launcher is pulling in a dep of an unstable version of openssl.

I feel like I've spent enough time in the build system and studio code to understand _what_ is happening but I have a couple of questions?

Is it expected behavior that our users would _all_ be using unstable builds of the launcher by default? If so, can someone share the reason why?

Another question is: because we release a fresh build of the launcher and it is unstable, does that obligate that build to be performed against other unstable packages or can it be built against stable instead? E.g. even if we're letting users ride the lightning of unstable, should we minimize some of these types of issues by obligating that _unstable thing_ to be built only against stable packages?

**Comment by christophermaier** [comment]
@eeyun I think the solution (for now) is as mentioned above: build packages using `hab studio build`, or enter an interactive studio with `HAB_STUDIO_SUP=false`. 

We do need to address this problem for real, though.

**Comment by eeyun** [comment]
Yeah I mean having a work around is good enough for now, but I feel like this points to possible concerns around our process for build/release?

Either way, its good to know that there are several options for getting around this and it's a trivial issue.

**Comment by stevendanna** [comment]
Does `HAB_FEAT_IGNORE_LOCAL=true` partially address this?
